std/assembly/staticarray.ts

```diff
@@ -284,27 +267,49 @@ export class StaticArray<T> {
     return out;
   }
 
-  slice(start: i32 = 0, end: i32 = i32.MAX_VALUE): Array<T> {
+  slice<U extends ArrayLike<T> = Array<T>>(start: i32 = 0, end: i32 = i32.MAX_VALUE): U {
     var length = this.length;
     start = start < 0 ? max(start + length, 0) : min(start, length);
-    end   = end   < 0 ? max(end   + length, 0) : min(end  , length);
+    end   = end   < 0 ? max(end   + length, 0) : min(end,   length);
     length = max(end - start, 0);
-    var slice = changetype<Array<T>>(__newArray(length, alignof<T>(), idof<Array<T>>()));
-    var sliceBase = slice.dataStart;
-    var thisBase = changetype<usize>(this) + (<usize>start << alignof<T>());
-    if (isManaged<T>()) {
-      let off = <usize>0;
-      let end = <usize>length << alignof<usize>();
-      while (off < end) {
-        let ref = load<usize>(thisBase + off);
-        store<usize>(sliceBase + off, ref);
-        __link(changetype<usize>(slice), ref, true);
-        off += sizeof<usize>();
+
+    var sourceBase = changetype<usize>(this) + (<usize>start << alignof<T>());
+    var size = <usize>length << alignof<T>();
+    let res!: U;
+
+    if (isArray<U>()) {
+      // return Array
+      res = changetype<U>(__newArray(length, alignof<T>(), idof<Array<T>>()));
+      let targetBase = changetype<Array<T>>(res).dataStart;
+      if (isManaged<T>()) {
+        let off: usize = 0;
+        while (off < size) {
+          let ref = load<usize>(sourceBase + off);
+          store<usize>(targetBase + off, ref);
+          __link(changetype<usize>(res), ref, true);
+          off += sizeof<usize>();
+        }
+      } else {
+        memory.copy(targetBase, sourceBase, size);
       }
     } else {
-      memory.copy(sliceBase, thisBase, length << alignof<T>());
+      assert(res instanceof StaticArray<T>);
```

Comment:
Can this be a compile time `ERROR`? Looks like the check must be statically true anyway?

Reply:
yeah